packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp

```diff
@@ -4437,10 +4445,15 @@ namespace Tpetra {
     }
     else {
       using LO = local_ordinal_type;
-      // Make sure that the GPU can see any updates made on host.
-      // This code only reads the local graph, so we don't need a
-      // fence afterwards.
-      execution_space().fence ();
+
+      // FENCE REVIEW - NOW PASSING
+      //   Testing: This code is exercised by unit tests. TpetraCore_CrsGraph_UnitTests0_MPI_4
+      //   GTX960:  Passed with CUDA_LAUNCH_BLOCKING=0 and fence removed.
+      //   White:   Passed with CUDA_LAUNCH_BLOCKING=0 and fence removed.
+      //   Plan:    Remove this fence.
+      //   Notes:   maxDifference looks properly setup as device.
+
+      // execution_space().fence ();
```

Comment:
@kddevin This one is a bit different from the above. Conservatively thinking, we also need the fence here as lclGraph_.row_map is on UVM and it can be still updated. In practice, the local graph construction should be done before this is called. This routine is computeLocalConstants. If we want a better code, we need a flag stating that the local graph is completed and safe to use it without fence. 

For now, I recommend to put the fence. Defining the state of local object can be improvements in the code.

Reply:
Updated comments #7546

Reply:
@kyungjoo-kim @kddevin I was looking into the possibility of adding a flag for this. I think we are talking about a bool bNeedFence in Tpetra_Graph_decl.hpp which is reset when needed (whenever we create the static graph). If computeLocalConstants is called again and the graph has not changed, we can skip the fence, similar to DualView need_sync_host behaviors. However it looks like there is no current test where adding such a flag will make a difference compared to just leaving the fence. The graph is always recreated before computeLocalConstants is called again. So I wanted to confirm if this is the idea you had in mind.

I did note 2 places where computeLocalConstants is used in Tpetra but not tested by a unit test.
    Call of computeLocalConstants in Tpetra_CrsMatrix_def.hpp.    line 4857
    Call of computeLocalConstants in graph fillComplete in Tpetra_CrsGraph_def.hpp.    line 3628

Reply:
@kyungjoo-kim @kddevin I may be figuring this out. I will plan to open a PR to clarify any questions I have.